### DISCUSSION TITLE
OpenTelemetry Instrumentation Support

### DISCUSSION THREAD

**Discussion by akoken** [ZiggyCreatures/FusionCache#153] [Ideas]
Hello, @jodydonetti!

First and foremost, I wanted to express my gratitude for the fantastic library you've created. Our microservices have been benefiting immensely from FusionCache, and we couldn't be happier with its performance.

I wanted to discuss a potential enhancement that could further amplify the observability of our systems. As you may be aware, "Tracing" is a vital aspect of system observability, and we currently rely on OpenTelemetry and Jaeger for distributed tracing. In light of this, I was wondering if you would consider adding FusionCache instrumentation for OpenTelemetry. Having such a feature would be truly remarkable and greatly enhance our ability to monitor and analyze the performance of FusionCache within our ecosystem.

Once again, thank you for your hard work and for creating such a valuable tool. We truly appreciate your dedication to the community. 

Thanks
Alp

**Comment by jodydonetti** [comment]
Hi @akoken and thanks for using FusionCache!

> First and foremost, I wanted to express my gratitude for the fantastic library you've created.
> [...]
> Once again, thank you for your hard work and for creating such a valuable tool. We truly appreciate your dedication to the community.

Thank you Abdurrahman, and I really mean it.
I work on FusionCache because it's something I deeply enjoy doing, and because I have the hope it may be of help for someone somewhere out there.
But then, when some kind soul like you express something like what you just said, it makes my day, and it's the cherry on top.
So thank you for your kind words.

> Our microservices have been benefiting immensely from FusionCache, and we couldn't be happier with its performance.

I'm glad it's being of help, thanks for using it!

> I wanted to discuss a potential enhancement that could further amplify the observability of our systems. As you may be aware, "Tracing" is a vital aspect of system observability, and we currently rely on OpenTelemetry and Jaeger for distributed tracing. In light of this, I was wondering if you would consider adding FusionCache instrumentation for OpenTelemetry. Having such a feature would be truly remarkable and greatly enhance our ability to monitor and analyze the performance of FusionCache within our ecosystem.

First thing: I'm no expert in OpenTelemetry but I would like to point out the work done by @JoeShook some time ago via third-party FusionCache plugins you can find linked in the repo's homepage:

![image](https://github.com/ZiggyCreatures/FusionCache/assets/1010086/0549f155-107e-4b4e-b354-24eb72b41a08)

Maybe they can be of some help? I don't know if they are (or can be in some way) used with OpenTelemetry.

Anyway, some time has passed since when those plugins have been created, and today OpenTelemetry has reached v1.0 (like FusionCache will, in a very short amount of time 😬).

Now, if you can help me jumpstart my OpenTelemetry knowledge a little bit, I may add _native support_ to FusionCache for that, officially, and right on time for the big v1.0 release.

Ideally I would need an existing sample of how a .NET library can integrate with OpenTelemetry to emit data: if you can point me to it I can do the rest with the specific knowledge of FusionCache.

What do you think?

Thanks!

**Comment by akoken** [comment]
Hi @jodydonetti,

Thank you for your quick response! Your dedication to FusionCache is truly appreciated.

Regarding the existing plugins by @JoeShook, I must admit that I didn't thoroughly investigate the source code and assumed they were more focused on metrics rather than tracing.

There is a contribution guide [here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/extending-the-sdk/README.md#instrumentation-library). While investigating, I came across the [StackExchangeRedis](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.StackExchangeRedis) instrumentation, which could serve as a good starting point. However, I noticed that using the `ActivitySource` provided in `System.Diagnostics.DiagnosticSource` might be a simpler approach, as outlined in the [OpenTelemetry .NET Tracing API documentation](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/README.md#introduction-to-opentelemetry-net-tracing-api). As far as I know, [MassTransit 8](https://github.com/MassTransit/MassTransit/discussions/2173) uses `ActivitySource` and doesn't need any seperate instrumentation package. 

Considering the potential impact on performance, it would be advisable to make this feature opt-in. Consumers can then evaluate the tradeoff and decide whether to enable it based on their specific requirements.

Once again, thank you for considering this enhancement. I'm excited about the possibilities it holds for further improving the observability of FusionCache. If there's anything I can do to assist or provide more information, please let me know.

Thanks!

**Reply by jodydonetti** [reply]
Awesome, thanks for the heads up!

Will look into it soon and let you know.

**Comment by JoeShook** [comment]
Nice to see someone interested in OTEL here.
As you can see I did publish a package,
JoeShook.ZiggyCreatures.FusionCache.Metrics.OpenTelemetry
<https://www.nuget.org/packages/JoeShook.ZiggyCreatures.FusionCache.Metrics.OpenTelemetry/1.0.4-alph003>
I focused on Metrics because a group I work with needed metrics with a
first class cache.  I have always wanted OTEL to be the implementation but
at the time OTEL in metrics was not ready.  While the OpenTelemetry package
is labeled as alpha it really isn't rocket science.  It is unit tested to
the same level as the other packages.

I think even without any work done at the FushionCache code base you will
still get great value from OTEL.  The package I created would correlate to
any OTEL tracing that exists already.  Meaning if you enabled OTEL in a
DotNet app and add my package you can trace a slow counter back to its
source.  Of course there might be some places in FusionChache where you
would want to call StartActivity.  I have a example built a while back here:
ZiggyCreatures.FusionCache.Metrics/examples/openTelemetry at main ·
JoeShook/ZiggyCreatures.FusionCache.Metrics · GitHub
<https://github.com/JoeShook/ZiggyCreatures.FusionCache.Metrics/tree/main/examples/openTelemetry>

I am a bit rusty on this and not putting a lot of time into this response.
But here are some more thoughts and comments...


   - Generally I try to avoid adding ActivitySources until I look at a real
   example app with traces.  I find there is a tone of trace data already.
   And if for example Redis is a backing store then I would be looking to use
   the StackeExchange.Redis
   <https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Instrumentation.StackExchangeRedis>
contrib
   package.  But again you might want to propogate tags.
   - Tags:  I have a core set of semantics in
   Plugins.Metrics.Core.SemanticConventions
   <https://github.com/JoeShook/ZiggyCreatures.FusionCache.Metrics/blob/main/src/FusionCache.Plugins.Metrics.Core/SemanticConventions.cs#L15>
   class describing the metadata such as Tags.  I think if we were to add
   extra tracing to FusionCache we should leverage this common semantics
   package.  It is the same technique you will find inside the DotNet OTEL SDK.
   - After chatting with the OTEL Slack group, the next steps of my
   OpenTelementry package would be to pick it up and put it in the
   opentelemetry-dotnet-contrib
   <https://github.com/open-telemetry/opentelemetry-dotnet-contrib> repo.
   -

Sorry, I can't get too deep on this at the moment.  But I wanted to let
everyone know I am thinking about it a little now, and could get sucked
back in.  :)

**Comment by jodydonetti** [comment]
Today.

![image](https://github.com/ZiggyCreatures/FusionCache/assets/1010086/af4b663a-1d40-4826-a6e5-5186f190c589)

**Reply by jodydonetti** [reply]
Hi @akoken and @JoeShook : it's taking shape, and it seems spectacular 😬

![image](https://github.com/ZiggyCreatures/FusionCache/assets/1010086/5e4a96cd-3271-48a0-ae27-3c9ee0cf5971)

Can't wait for this to be released.

Oh, and the sample in the screenshot is using [fail-safe](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/FailSafe.md), some simulated delays (via chaos components), [soft/hard timeouts](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/Timeouts.md), background factory completion, background execution of [distributed cache](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/CacheLevels.md) and [backplane](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/Backplane.md) and more, all nicely rendered in the Jaeger UI.

**Reply by jodydonetti** [reply]
And here it is the [official issue](https://github.com/ZiggyCreatures/FusionCache/issues/187) to track design and development: if you have some time please read it and let me know what you think.

Thanks!

**Reply by JoeShook** [reply]
Wow Jody, this is good to hear!

Looks like I have to catch up on what is happening.  FYI I did release an OpenTelemetry package by the name of [JoeShook.ZiggyCreatures.FusionCache.Metrics.OpenTelemetry](https://www.nuget.org/packages/JoeShook.ZiggyCreatures.FusionCache.Metrics.OpenTelemetry).  Seems my main README needs updating.
I will need to visit your official issue and catch up.  The good thing is I may be refocusing on metrics for work reasons again and can give this more focus.

@jodydonetti as I quickly think about this, there is one thing I want to make sure you capture as you look to add OTEL tracing.  That is a naming conventions and the ability of a package consumer to set the naming convention.   This is facilitated via an implementation of the ```ISemanticConventions``` that is passed into the ```MetricsMeter```.  By default the metrics packages include a default ```ISemanticConventions```.   When I wrote these pacakges I was taking my influence of the "class of semantics" from the official dotnet otel [source code.](https://github.com/open-telemetry/opentelemetry-dotnet/blob/13791e268e9ca6744f64c74951b718a8c7002bcc/src/Shared/SemanticConventions.cs#L13).  Actually there is more than just ```ISemanticConventions``` needed to control metrics naming conventions.  Just take a look at the [FusionMeter constructor](https://github.com/JoeShook/ZiggyCreatures.FusionCache.Metrics/blob/70cad76906d4c719038c028e6941ee6c29a630f3/src/FusionCache.Plugins.Metrics.OpenTelemetry/FusionMeter.cs#L59) and you can see it in action.  

- Joe

**Reply by jodydonetti** [reply]
Hi Joe, thanks for your inputs, will look into them!

**Reply by jodydonetti** [reply]
Anybody have been able to try out the `v0.25.0-preview1`?

Any opinion will be greatly appreciated!

**Reply by jodydonetti** [reply]
Hi all, I just release [v0.25.0](https://github.com/ZiggyCreatures/FusionCache/releases/tag/v0.25.0) 🥳

**Comment by jodydonetti** [comment]
Hi @akoken , I just release [v0.25.0](https://github.com/ZiggyCreatures/FusionCache/releases/tag/v0.25.0) which contains full observability support.

All 3 signals are included:
- logs (always have been)
- traces (new)
- metrics (new)

I also added some official [documentation](https://github.com/ZiggyCreatures/FusionCache/blob/main/docs/OpenTelemetry.md).

Hope this helps, and please mark this as the answer if you feel like it.
